src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -138,6 +140,41 @@ double JacobianPerformanceTest(ModuleType& module,
   return arma::max(arma::max(arma::abs(centralDifference - delta)));
 }
 
+// Simple numerical gradient checker.
+template<class FunctionType>
+double CheckGradient(FunctionType& function, const double eps = 1e-7)
+{
+  // Get gradients for the current parameters.
+  arma::mat orgGradient, gradient, estGradient;
+  function.Gradient(orgGradient);
+
+  estGradient = arma::zeros(orgGradient.n_rows, orgGradient.n_cols);
+
+  // compute numeric approximations to gradient.
+  for (size_t i = 0; i < orgGradient.n_elem; ++i)
+  {
+    double tmp = function.Parameters()(i);
+
+    // Perturb parameter with a positive constant and get costs.
+    function.Parameters()(i) += eps;
+    double costPlus = function.Gradient(gradient);
+
+    // Perturb parameter with a negative constant and get costs.
+    function.Parameters()(i) -= (2 * eps);
+    double costMinus = function.Gradient(gradient);
+
+    // Restore the parameter value.
+    function.Parameters()(i) = tmp;
+
+    // Compute numerical gradients using the costs calculated above.
+    estGradient(i) = (costPlus - costMinus) / (2 * eps);
+  }
```

Comment:
(This is a little off-topic but maybe a bit useful.)  This is the "finite differences stochastic approximation" (FDSA) of the gradient, which is a useful technique.  But sometimes it can be a bit slow if there are a lot of dimensions.  You can get a noisier but quicker approximating by using "simultaneous perturbation stochastic approximation" (SPSA), where instead of modifying just one parameter at a time you modify all parameters simultaneously with a random perturbation vector:

https://en.wikipedia.org/wiki/Simultaneous_perturbation_stochastic_approximation

Anyway I doubt that the runtime here is actually an issue, I just wanted to point it out because I think SPSA is a cool technique.